### PR TITLE
[NEW] #8771 - fine granular settings permissions

### LINKED ISSUE #8771: [Feature request] More fine grained permission system
### Description:

Currently, rocket chat does not allow giving permissions for subcategories of "Settings". This leads to some issues, if, e.g. one wants to allow someone to edit the main page or change the CSS but does not want to give them access to the LDAP settings (since they contain sensitive data).

Instead, one should be able to set permissions for the individual subcategories.

### PR DESCRIPTION
**Pull request by mrsimpson**
Closes #8771

@RocketChat/core This PR adds the ability to allow a role to change only dedicated settings.
This is necessary in order to e. g. allow a user to change the colors of his Rocket.Chat without compromising critical settings such as LDAP-connectivity.

Though the actual purpose is already fulfilled, this PR has to be considered WIP due to the limitations mentioned below.

**@Developers (particularly frontend-developers): I need assistance to fix the "final" (mostly UI) bugs**

## What works
- [x] Permissions for each and every setting is created automatically
- [x] The permissions-UI is extended to visualize and manage all setting-permissions just like normal permissions.
- [x] The settings-UI in the admin-overview displays the selected settings for the role
- [x] setting-based permissions are respected/validated when changing a setting
- [x] permissions are reactive 

## What could be improved

- In general, the permissions-UI is not beautiful. This is due to the huge width of the permissions table. Any help on that is appreciated - though @engelgabriel said you'd redesign the look anyway. Currently, there's a CSS-hack which does not work in all browsers to make the header take the full width. There has to be another option, but I'm not a CSS-native.

## Impressions

Configuring the permissions

setting-permissions

There is one explicit permission which makes the whole set applicable.
Whether this is a good idea, I am not convinced myself. The permission could be determined.

selected-settings

Effect

effect reduced settings

### CODE REVIEW

**Comment by gdelavald** [comment]
This is great, thanks for the PR. We'll check if we have people available to help on the needed changes, else I'm sure the community can help improving it

File: package.json [hunk 1]
```diff
@@ -60,7 +60,9 @@
 		"chat"
 	],
 	"scripts": {
-		"start": "meteor npm i && meteor",
+		"start": "meteor npm i && meteor run",
+		"debug": "meteor run --inspect",
```
Comment (mrsimpson) [inline package.json]: This was actually independent of this PR, but somehow made its way into it. It does no harm. Please remove if not favored.

File: packages/rocketchat-authorization/client/stylesheets/permissions.css [hunk 2]
```diff
@@ -9,6 +9,10 @@
 		font-weight: bold !important;
 	}
 
+	& .section:not(.section-collapsed) {
+		inline-size: fit-content;
```
Comment (mrsimpson) [inline packages/rocketchat-authorization/client/stylesheets/permissions.css]: This is the hack: I didn't get the section headline to take the full available width otherwise.
There's surely a better (all-browser-compatible) solution to that.

File: packages/rocketchat-authorization/client/views/permissions.html [hunk 3]
```diff
@@ -1,36 +1,76 @@
+<template name="permissionsTable">
```
Comment (mrsimpson) [inline packages/rocketchat-authorization/client/views/permissions.html]: Created a template for the table as both permission tables follow the same scheme.

File: packages/rocketchat-authorization/client/views/permissions.js [hunk 4]
```diff
@@ -1,66 +1,121 @@
 /* globals ChatPermissions */
+import {permissionLevel} from '../../lib/rocketchat';
+
+const whereNotSetting = {
```
Comment (mrsimpson) [inline packages/rocketchat-authorization/client/views/permissions.js]: `$ne` is not supported by Minimongo. Thus, had to fallback to `$where`. Still, this does not seem to work, see the `.filter`-comment some lines below

File: packages/rocketchat-authorization/client/views/permissions.js [hunk 5]
```diff
@@ -1,66 +1,121 @@
...
+	roles() {
 		return Template.instance().roles.get();
 	},
 
-	permission() {
-		return ChatPermissions.find({}, {
-			sort: {
-				_id: 1
+	permissions() {
+		return ChatPermissions.find(whereNotSetting, //the $where seems to have no effect - filtered as workaround after fetch()
```
Comment (mrsimpson) [inline packages/rocketchat-authorization/client/views/permissions.js]: any idea why the `$where` has no effect?

File: packages/rocketchat-authorization/lib/rocketchat.js [hunk 6]
```diff
@@ -1 +1,5 @@
 RocketChat.authz = {};
+
+export const permissionLevel = {
+	SETTING: 'setting'
```
Comment (mrsimpson) [inline packages/rocketchat-authorization/lib/rocketchat.js]: I can imagine that there are other levels as well, such as the permission to change dedicated roles, thus making it a map

File: packages/rocketchat-authorization/server/publications/permissions.js [hunk 7]
```diff
@@ -20,4 +27,12 @@ Meteor.methods({
 
 RocketChat.models.Permissions.on('changed', (type, permission) => {
 	RocketChat.Notifications.notifyLoggedInThisInstance('permissions-changed', type, permission);
```
Comment (mrsimpson) [inline packages/rocketchat-authorization/server/publications/permissions.js]: Should this not be `notifyLogged()` it was like that earlier, so I didn't change it...

File: packages/rocketchat-ui-admin/client/SettingsCachedCollection.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+export class PrivateSettingsCachedCollection extends RocketChat.CachedCollection {
+	constructor() {
+		super({
+			name: 'private-settings',
+			eventType: 'onLogged'
+		});
+	}
+
+	setupListener(eventType, eventName) {
```
Comment (mrsimpson) [inline packages/rocketchat-ui-admin/client/SettingsCachedCollection.js]: The redefinition is necessary as the change of permissions has to trigger the Settings-collection to be refreshed. We'll have to listen to an event "authorization changed".

File: packages/rocketchat-ui-admin/client/SettingsCachedCollection.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+		super.setupListener(eventType, eventName);
+
+		// private settings also need to listen to a change of authorizationsfor the setting-based authorizations
+		RocketChat.Notifications[eventType || this.eventType](eventName || this.eventName, (t, record) => {
+			this.log('record received', t, record);
+			if (t === 'auth') {
```
Comment (mrsimpson) [inline packages/rocketchat-ui-admin/client/SettingsCachedCollection.js]: If there was an enumeration-constant, I'd be happy to use it ;)

File: packages/rocketchat-ui-admin/client/SettingsCachedCollection.js [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+				if (! (RocketChat.authz.hasAllPermission([`change-setting-${ record._id }`, 'manage-selected-settings'])
```
Comment (mrsimpson) [inline packages/rocketchat-ui-admin/client/SettingsCachedCollection.js]: The concatenation `change-setting-${ record._id }` should be reusable function of the auth-package. However, as `rocketchat:lib` is lower than `rocketchat:auth`, this is not possible.
For simplicity (and worse maintainability) I just green-coded that...

**Comment by mrsimpson** [comment]
removed WIP as this is actually done, except beautification

**Comment by Wirone** [comment]
Is there possibility to merge this feature in nearest future?

**Comment by mrsimpson** [comment]
@RocketChat/core I merged the current `master` in order to allow other users who build their own version to merge this branch. If you could let me know what you need to get it merged (and when!), I'll merge develop.

**Comment by mrsimpson** [comment]
@RocketChat/core 
I updated the PR with the latest develop and resolved the conflicts. Automated tests are also available. Would you mind re-reviewing this PR?

File: tests/end-to-end/ui/13a-setting-permissions.js [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+describe('[Rocket.Chat Settings based permissions]', function() {
+
+	const newTitle = 'Testtitle';
+
+	describe('Give User Permissions', function() {
+		before(() => {
+			try {
+				//If the tests run as a suite,a user may already be logged-in
```
[the remark this answers is not in the document]
Comment (mrsimpson) [reply tests/end-to-end/ui/13a-setting-permissions.js]: now that is picky.

**Comment by mrsimpson** [comment]
@geekgonecrazy now, merged `develop` - even more boring.

File: .scripts/seperateTesting.sh [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    set +x
+    if [ $failed -eq 0 ]; then
+      break
+    fi
+  done
+  if [ $failed -ne 0 ]; then
+    exit 1
+  fi
+  rm $tmpPath/${file##*/}
+done
```
Comment (geekgonecrazy) [inline .scripts/seperateTesting.sh]: I wonder if this change in testing should be in a different PR?

Comment (mrsimpson) [reply .scripts/seperateTesting.sh]: Of course it could be. I left it in here since it makes it easier to validate the tests added ;-)
But of course, this is useful even without the actual feature

File: app/lib/server/methods/saveSettings.js [hunk 13]
```diff
@@ -41,6 +40,14 @@ Meteor.methods({
 			}
 			settings.updateById(_id, value, editor);
 		});
+		// Throw messages for settings that are not allowed to save!
+		if (settingsNotAllowed.length) {
+			throw new Meteor.Error('error-action-not-allowed', 'Editing settings is not allowed', {
```
Comment (mrsimpson) [inline app/lib/server/methods/saveSettings.js]: This error is not propagated to the user

File: app/lib/server/methods/saveSettings.js [hunk 14]
```diff
@@ -1,27 +1,26 @@
...
 	saveSettings(params = []) {
+		const settingsNotAllowed = [];
 		if (Meteor.userId() === null) {
 			throw new Meteor.Error('error-action-not-allowed', 'Editing settings is not allowed', {
 				method: 'saveSetting',
 			});
 		}
 
-		if (!hasPermission(Meteor.userId(), 'edit-privileged-setting')) {
-			throw new Meteor.Error('error-action-not-allowed', 'Editing settings is not allowed', {
```
Comment (mrsimpson) [inline app/lib/server/methods/saveSettings.js]: This actually failed the saving of multiple settings on the first permission violation. Since now multiple settings can be saved at once and some may fail while others wont, I moved this error handling into a loop down below.

File: app/ui-admin/client/SettingsCachedCollection.js [hunk 15]
```diff
@@ -0,0 +1,38 @@
+import _ from 'underscore';
```
Comment (mrsimpson) [inline app/ui-admin/client/SettingsCachedCollection.js]: The setting collection was used twice, so I refactored this into an own file.

File: packages/rocketchat-authorization/client/stylesheets/permissions.css (hunk 2 above)
Comment (mrsimpson) [reply packages/rocketchat-authorization/client/stylesheets/permissions.css]: not needed anymore - switched to a second table

File: package.json (hunk 1 above)
Comment (mrsimpson) [reply package.json]: removed

**Comment by ggazzo** [comment]
image

**Comment by ggazzo** [comment]
@sampaiodiego could you review this code plz? I'm not worthy anymore =/ =) ...

ps: last thing I know, we have to move the streamer to `emitter.js` AND check if its everything ok because we gonna be sending the settings changes for all "manage-settings" users
